File: apps/web/lib/api/approvals.ts
import { apiFetch } from "./client";

export interface ApprovalItem {
  id: string;
  workflow_id: string;
  action_type: string;
  tool_name: string;
  parameters: any;
  reason?: string;
  status: string; // pending, approved, rejected
  risk_level?: string;
  payload?: any;
  created_at: string;
}

export async function listApprovals(): Promise<ApprovalItem[]> {
  try {
    return await apiFetch<ApprovalItem[]>("/api/approvals");
  } catch {
    return [];
  }
}

export async function getPendingApprovals(): Promise<ApprovalItem[]> {
  try {
    const all = await listApprovals();
    return all.filter((a) => a.status === "pending" || a.status === "awaiting_approval");
  } catch {
    return [];
  }
}

export async function decideApproval(id: string, decision: "approved" | "rejected", comment?: string) {
  return apiFetch(`/api/approvals/${id}/decision`, {
    method: "POST",
    body: JSON.stringify({ decision, comment }),
  });
}

export async function resolveApproval(id: string, status: "approved" | "rejected", comment?: string) {
  return decideApproval(id, status, comment);
}

export const approvalsApi = {
  listApprovals,
  getPendingApprovals,
  decideApproval,
  resolveApproval,
};

File: apps/api/app/api/routes/approvals.py
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field
from sqlalchemy import select, desc
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db, User, Approval, Workflow
from app.api.deps import get_current_user
from app.services.workflow_service import workflow_service

router = APIRouter(prefix="/approvals", tags=["Approvals"])


class ApprovalActionRequest(BaseModel):
    decision: str = Field(..., description="'approved' or 'rejected'")
    comment: Optional[str] = None


@router.get("")
async def list_pending_approvals(
    status_filter: str = "pending",
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """List pending human approvals requiring user intervention."""
    stmt = (
        select(Approval)
        .join(Workflow, Approval.workflow_id == Workflow.id)
        .where(Workflow.user_id == user.id)
    )
    if status_filter:
        stmt = stmt.where(Approval.status == status_filter)
    stmt = stmt.order_by(desc(Approval.created_at))

    res = await db.execute(stmt)
    approvals = res.scalars().all()
    return [
        {
            "id": a.id,
            "workflow_id": a.workflow_id,
            "step_id": a.step_id,
            "action_name": a.action_name,
            "action_payload": a.action_payload,
            "status": a.status,
            "created_at": a.created_at.isoformat() if a.created_at else None,
            "reviewed_at": a.reviewed_at.isoformat() if a.reviewed_at else None,
            "comments": a.comments,
        }
        for a in approvals
    ]


@router.post("/{approval_id}/decision")
async def decide_approval(
    approval_id: str,
    payload: ApprovalActionRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Approve or reject a pending sensitive action, unpausing the workflow if approved."""
    stmt = (
        select(Approval)
        .join(Workflow, Approval.workflow_id == Workflow.id)
        .where(Approval.id == approval_id, Workflow.user_id == user.id)
    )
    res = await db.execute(stmt)
    apprv = res.scalar_one_or_none()
    if not apprv:
        raise HTTPException(status_code=404, detail="Approval request not found")

    success = await workflow_service.resolve_approval(
        workflow_id=apprv.workflow_id,
        approval_id=approval_id,
        decision=payload.decision,
        comment=payload.comment
    )
    return {
        "approval_id": approval_id,
        "status": payload.decision,
        "workflow_id": apprv.workflow_id,
        "resumed": payload.decision.lower() in ["approved", "approve"] and success
    }
